### ISSUE TITLE
Hard crash with std::bad_alloc while executing query

### ISSUE THREAD

**Issue by bburket** [arangodb/arangodb#3458]
## my environment running ArangoDB
I'm using the latest ArangoDB of the respective release series:
3.5 

Mode:
- Single Server
Storage-Engine:
- [rocksdb

On this operating system:
AWS Linux (CentOS)

### this is an AQL-related issue:
[ x] I'm using graph features

I'm issuing AQL via:
- [x ] web interface with this browser:     running on this OS:

When executing this query against apx 30 mil docs in Table1 and 20 mil docs in Table2 I get a hard crash with std::bad_alloc. Server has apx 32gb of RAM.

Everytime I execute this the EdgeTable collection will get exactly 1000000 rows written to it before the server hard crashes

FOR p in Table1
   FOR s in Table2
   FILTER TO_STRING(p.someId) == s._key
   INSERT {
       _key: CONCAT(s._key, "_", p._key) ,
       _from: p._id,
       _to: s._id
   } INTO EdgeTable OPTIONS { ignoreErrors: true }

**Comment by bburket** [comment]
Correct. 3.4.5

https://github.com/arangodb/arangodb/issues/3342

I tried a bunch of things but this actually prevented OOMs from happening. I suspect this was the issue. I have not had a crash since I made this change and have committed several large transactions (10s of millions of rows)

**Comment by jsteemann** [comment]
Ok, thanks. By the way, we are currently investigating several memory-related issues and have some clues for how to improve the situation in one of the next releases.

**Comment by dchusovitin** [comment]
Same error on version 3.2.5 (from docker hub) with rocksdb storage engine

**Comment by dchusovitin** [comment]
And exception with running query (without crash) on 3.2.1
std::bad_alloc (exception location: /var/lib/otherjenkins/workspace/RELEASE__BuildPackages/arangod/GeneralServer/RestHandler.cpp:212)

**Comment by dothebart** [comment]
Hi, 
with ArangoDB 3.2.8 we fixed some important memory management issues by applying upstream V8 patches and improving our jemalloc integration.

Can you please retry whether this is still an issue after the upgrade?

**Comment by dothebart** [comment]
Closing as fixed with 3.2.8

**Comment by lewisdiamond** [comment]
I get this in Arangodb 3.3.2
Rocksdb
With the following system config:
```
sudo sysctl -w "vm.max_map_count=512000
sudo bash -c "echo madvise > /sys/kernel/mm/transparent_hugepage/enabled"
sudo bash -c "echo madvise > /sys/kernel/mm/transparent_hugepage/defrag"
sudo bash -c "echo 2 > /proc/sys/vm/overcommit_memory"
sudo bash -c "echo 100 > /proc/sys/vm/overcommit_ratio"
```
```
terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc
```

**Comment by lewisdiamond** [comment]
Related to #3889

**Comment by tomterl** [comment]
I hope this is indeed related and a helpful detail: 
I switched an arangod testing instance to 3.3.3 with rocksdb, from 3.2.x with mmfiles.  I have a similar problem (std::bad_alloc during arangorestore) seemingly connected to the 'overcommit_memory' setting. If I set it to '2' as recommended, the dump in question reliably crashes the recovery-process; set to '0' the restore succeeds.  This is on Ubuntu 16.04, kernel 4.4.0-109-generic; restore and server running on the same machine.

**Comment by tomterl** [comment]
I committed a bit of time to what's going on under the hood, setting overcommit to '2' should be the right thing to do. Maybe in our case additional tuning of the ratio would be necessary to avoid the crash, or simply putting more memory in the machine; but since the machine is perfectly capable of handling the workload in case of the more optimistic allocation heuristic, this is perceived as an arandodb problem. We will have a bigger machine with a lot of ram allocated to arangodb one of these days and I will try to import the dump there, see how that goes, and report back.

If this is perceived as hijacking this issue, please tell me so and don't be silently annoyed :-) , it obviously looked like the right place to me.

**Comment by dothebart** [comment]
can you have a word on your machine dimensions and the dump size (number collections, number documents, indices, average document size)?  Can you share your dataset?

**Comment by tomterl** [comment]
Alas, as fortunate as it would be in a case like this, I cannot share the dataset.

The machine: Intel Core i5-3470 (4 cores, 4 threads); 16GB RAM; 120GB SSD; 
The dump is ~613M in total; 6 Collections, the biggest has ~460000 documents with an average size of  ~900B (taking up rougly 2/3 of the dump) and a lot of indices defined. I took a look at the dump-files: 

    collection | size | count | avg | indices
    -----------+------+-------+-----+--------
        1      | 396M | 460000| 900B| 75
        2      | 147M | 460000| 330B|  2
        3      |  27M |  54700| 520B|  7
        4      |  25M | 125000| 210B|  3
        5      |  20M |  69300| 300B|  6

Nothing out of the ordinary; maybe with the exception of the indices count for the first collection. I'll experiment with removing indices from the first collection (I assume I can edit the XXX-structure.json files accordingly), see what happens.

**Comment by dothebart** [comment]
yes, you should be able to do that. 
If you are able to narrow this down, it would be great if you can share a `lorem ipsem` dataset with us reproducing the issue

**Comment by vinaypyati** [comment]
Hi @tomterl,

Just sending a quick follow up. Did you have a chance to reproduce this issue on the latest ArangoDB v3.3.16 or on v3.4.0 RC1. Please let us know your feedback. 

Thanks!

**Comment by tomterl** [comment]
Hi @vinaypyati, 

the work on this particular project is on hold ATM, and I'm unable to allocate time to test in the coming couple of weeks -- I'll report back as soon as I can with findings in our current arangodb-setup (machines are bigger, etc.).

**Comment by vinaypyati** [comment]
Hi @tomterl,

Just a quick follow up. How things are going?
3.4.0 RC3 has been released. You can download it from the [download section](https://www.arangodb.com/download-technical-preview/). Your feedback is highly appreciated. 

Thanks!

**Comment by tomterl** [comment]
Hi @vinaypyati ,

I did not install RC3, but restoring into a current 3.4-devel does not crash the server. It takes a really long time (~29min) and fills the log with 'slow background settings sync:...' messages -- but it doesn't crash. I made sure overcommit_memory is set to 2 before starting the server.

The restore was overwriting an existing instance of the db -- that's a possible reason for the slowness. 

Cheers!

**Comment by maxkernbach** [comment]
Closing this issue as fixed since the crash does not occur using 3.4.0+.

However, please note that the recommended setting for [`overcommit_memory`](https://www.arangodb.com/docs/stable/installation-linux-osconfiguration.html#over-commit-memory) changed from **2** to  **0** or **1**.  For further details, please have a look at the **Linux System Configuration** section of the article below:

https://www.arangodb.com/2019/03/small-steps-reduce-arangodb-resource-footprint/
